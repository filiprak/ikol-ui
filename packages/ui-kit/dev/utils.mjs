import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
