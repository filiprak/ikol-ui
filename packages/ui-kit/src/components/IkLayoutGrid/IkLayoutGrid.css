.ik-layout-grid {
    position: relative;
    display: flex;
    flex: 1 1 auto;
    overflow: hidden;
    width: 100%;
}

.ik-layout-grid .ik-layout-grid-window:not(:last-of-type),
.ik-layout-grid-window--border {
    border-right: 1px solid var(--surface-darken-color);
}

.ik-layout-grid-item {
    display: flex;
    flex: 1 0 0;
    flex-direction: column;
    max-width: 100%;
    max-height: 100%;
}

.ik-layout-grid-item--row {
    flex-direction: row;
}

.ik-layout-grid-window {
    display: flex;
    align-items: stretch;
    flex-direction: column;
    overflow: auto;
    flex: 1 1 auto;
    max-width: 100%;
    max-height: 100%;
}

.ik-layout-grid-window--static {
    flex: 0 0 auto;
}

.ik-layout-grid-window__header {
    display: flex;
    flex: 0 0 auto;
    align-items: center;
    max-width: 100%;
    max-height: 100%;
    padding: 24px 25px 16px 25px;
    line-height: 32px;
}

.ik-layout-grid-window--dense .ik-layout-grid-window__header {
    padding: 15px;
    line-height: 24px;
}

.ik-layout-grid-window--no-header-pad .ik-layout-grid-window__header {
    padding: 0;
}

.ik-layout-grid-window__title {
    flex: 1 1 auto;
    font-size: 13px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    min-width: 0;
}

.ik-layout-grid-window__actions {
    flex: 0 0 auto;
    font-size: 18px;
}

.ik-layout-grid-window__actions .ik-button {
    padding: 4px 10px;
    border-radius: 20px;
}

.ik-layout-grid-window__content {
    position: relative;
    display: flex;
    flex-direction: column;
    align-items: stretch;
    flex: 1 1 auto;
    min-height: 0;
    overflow: hidden;
    max-width: 100%;
    max-height: 100%;
}

.ik-layout-grid-window__top,
.ik-layout-grid-window__bottom {
    flex: 0 0 auto;
    max-width: 100%;
}

.ik-layout-grid-window__action {
    display: inline-flex;
    align-items: center;
    font-size: 13px;
    margin-left: 2px;
    color: var(--on-background-color-3);
    cursor: pointer;
}

.ik-layout-grid-window__action:hover {
    color: var(--on-background-color-2);
}