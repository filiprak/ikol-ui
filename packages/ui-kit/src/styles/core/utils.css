/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
