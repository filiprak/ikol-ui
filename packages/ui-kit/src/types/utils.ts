import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
