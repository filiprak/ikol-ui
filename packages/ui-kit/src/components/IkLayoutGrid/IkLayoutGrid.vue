<template>
    <div class="ik-layout-grid">
        <slot></slot>
    </div>
</template>
<script setup lang="ts">
import '@ui/styles';
import './IkLayoutGrid.css';
</script>
