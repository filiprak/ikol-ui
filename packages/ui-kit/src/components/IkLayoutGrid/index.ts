import IkLayoutGrid from './IkLayoutGrid.vue';
import IkLayoutGridItem from './IkLayoutGridItem.vue';
import IkLayoutGridWindow from './IkLayoutGridWindow.vue';

export interface IkLayoutGrid extends InstanceType<typeof IkLayoutGrid> { };
export interface IkLayoutGridItem extends InstanceType<typeof IkLayoutGridItem> { };
export interface IkLayoutGridWindow extends InstanceType<typeof IkLayoutGridWindow> { };

export { IkLayoutGrid, IkLayoutGridItem, IkLayoutGridWindow };
