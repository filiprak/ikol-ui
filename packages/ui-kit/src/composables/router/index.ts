import type { InjectionKey, Ref } from 'vue';
import { inject } from 'vue';
import { getCurrentInstance, logError } from '@ui/utils/core';
import { IkRouterManager } from '@ui/composables/router/IkRouterManager';
import type { IkRouteManager } from '@ui/composables/router/IkRouteManager';
import type { RouteRecordRaw } from 'vue-router';
import { createRouter as _createRouter, createWebHashHistory } from 'vue-router';

export const ROUTER_SYMBOL: InjectionKey<IkRouterManager> = Symbol.for('ik-router');

export interface IkRouterOptions {
    routes: Readonly<RouteRecordRaw[]>
}

export function createRouter(options: IkRouterOptions): IkRouterManager {
    return new IkRouterManager(
        _createRouter({
            history: createWebHashHistory(),
            routes: options.routes,
        })
    );
}

export function useRouter(): IkRouterManager {
    let router;

    if (getCurrentInstance('useRouter')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router;
}

export function useRoute(): Ref<IkRouteManager> {
    let router;

    if (getCurrentInstance('useRoute')) {
        router = inject(ROUTER_SYMBOL);
    }

    if (!router) {
        router = createRouter({ routes: [] });
        logError('Injection "router" not found');
    }

    return router.route;
}
