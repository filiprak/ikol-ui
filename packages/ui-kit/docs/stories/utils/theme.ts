import { useTheme } from '@ui/composables';
import { createIkolUI } from '@ui/index';
import { App } from 'vue';
import { StoryContext, VueRenderer } from '@storybook/vue3';

export function setupTheme(context: any) {
    const theme = useTheme();

    theme.is_dark.value = context?.globals?.theme === 'dark';

    return theme;
}

export function setupApp(app: App, context?: StoryContext<VueRenderer>) {
    app.use(createIkolUI());
}
