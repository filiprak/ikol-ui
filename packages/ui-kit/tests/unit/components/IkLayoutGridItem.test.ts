import type { RenderOptions } from '@ui/tests/tools/utils';
import { renderShallow } from '@ui/tests/tools/utils';
import { IkLayoutGridItem } from '@ui/components/IkLayoutGrid';
import { getRouterMock } from '@ui/tests/mocks/router';

const defineRouter = () => {
    return getRouterMock({
        routes: [
            {
                path: '/',
                name: 'UnitTestRoute',
                component: {},
            },
            {
                path: '/other',
                name: 'UnitTestRouteOther',
                component: {},
            },
            {
                path: '/unit',
                name: 'UnitTestRouteUnit',
                component: {},
            },
        ],
    });
};

describe('IkLayoutGridItem', () => {
    let router = defineRouter();

    const getInstance = (options: RenderOptions<typeof IkLayoutGridItem> = {}) => renderShallow(IkLayoutGridItem, {
        router,
    }, options);

    beforeAll(async () => {
        router = defineRouter();

        await router.push('/');
    });

    it('renders with correct default state', () => {
        const wrapper = getInstance();

        expect(wrapper.classes()).toStrictEqual([
            'ik-layout-grid-item',
        ]);

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with default slot', () => {
        const wrapper = getInstance({
            slots: {
                default: 'This is default slot...',
            },
        });

        expect(wrapper.text()).toContain('This is default slot...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders as row', () => {
        const wrapper = getInstance({
            props: {
                row: true,
            },
        });

        expect(wrapper.classes()).toContain('ik-layout-grid-item--row');
    });

    it('renders with size', () => {
        const wrapper = getInstance({
            props: {
                size: 99,
            },
        });

        expect(wrapper.attributes('style')).toContain('flex-basis: 99%');
    });

    it('renders on mobile as full size', async () => {
        const wrapper = getInstance({
            props: {
                mobile_route: 'UnitTestRoute',
            },
            device: { mobile: true },
        });

        expect(wrapper.attributes('style')).toContain('flex-basis: 100%');
    });

    it('hides on mobile when does not match route', async () => {
        const wrapper = getInstance({
            props: {
                mobile_route: 'UnitTestRouteOther',
            },
            device: { mobile: true },
        });

        expect(wrapper.attributes('style')).toContain('display: none;');
    });

    it('shows on mobile when matches any of given routes', async () => {
        await router.replace('/unit');

        const wrapper = getInstance({
            props: {
                mobile_route: 'UnitTestRouteOther,UnitTestRouteUnit',
            },
            device: { mobile: true },
        });

        expect(wrapper.attributes('style')).not.toContain('display: none;');
    });
});
