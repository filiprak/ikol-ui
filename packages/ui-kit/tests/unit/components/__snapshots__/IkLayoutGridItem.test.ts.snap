// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkLayoutGridItem renders with correct default state 1`] = `
<div
  class="ik-layout-grid-item"
>
  
  
</div>
`;

exports[`IkLayoutGridItem renders with default slot 1`] = `
<div
  class="ik-layout-grid-item"
>
  
  This is default slot...
  
</div>
`;
